index.bs

```diff
@@ -3062,17 +3055,29 @@ To <dfn>remove attribution scopes for sources with destination</dfn> given a [=s
       * |a|[1] is greater than |b|[1].
       * |a|[1] is equal to |b|[1] and |a|[0] is greater than |b|[0].
 1. Let |selectedAttributionScopes| be a new [=set=].
-1. Let |sourcesToClearAttributionScopes| be a new [=set=].
+1. Let |sourcesToRemove| be a new [=set=].
 1. [=list/iterate|For each=] |record| of |attributionScopeRecords|:
     1. If |selectedAttributionScopes|'s [=set/size=] is less than |pendingSource|'s [=attribution source/attribution scope limit=], [=set/append=] |record|[0] to |selectedAttributionScopes|.
-    1. Otherwise, if |selectedAttributionScopes| does not [=set/contain=] |record|[0], [=set/append=] |record|[2] to |sourcesToClearAttributionScopes|.
+    1. Otherwise, if |selectedAttributionScopes| does not [=set/contain=] |record|[0], [=set/append=] |record|[2] to |sourcesToRemove|.
 1. [=set/iterate|For each=] |source| of the [=attribution source cache=]:
-    1. If |sourcesToClearAttributionScopes| [=set/contains=] |source|'s [=attribution source/source identifier=], set |source|'s [=attribution source/attribution scopes=] to a new [=set=].
+    1. If |sourcesToRemove| [=set/contains=] |source|'s [=attribution source/source identifier=], [=set/remove=] |source| from the [=attribution source cache=].
 
-To <dfn>remove attribution scopes for sources with attribution scopes not selected</dfn> given an [=attribution source=] |pendingSource|:
+To <dfn>remove sources with attribution scopes not selected</dfn> given an [=attribution source=] |pendingSource|:
 1. If |pendingSource|'s [=attribution source/attribution scope limit=] is null, return.
 1. [=set/iterate|For each=] |destination| in |pendingSource|'s [=attribution source/attribution destinations=]:
-    1. [=Remove attribution scopes for sources with destination=] with |destination| and |pendingSource|.
+    1. [=Remove sources with attribution scopes not selected for destination=] with |destination| and |pendingSource|.
+
+To <dfn>remove or update sources with incompatible attribution scope fields from cache</dfn> given an [=attribution source=] |pendingSource|:
+1. Let |matchingSources| be the result of running [=find sources with common destinations and reporting origin=] with |pendingSource|.
+1. [=list/iterate|For each=] |source| of |matchingSources|:
+    1. If |pendingSource|'s [=attribution source/attribution scopes=] [=set/is empty|empty=] and |source|'s [=attribution source/attribution scope limit=] is not null:
+        1. Set |source|'s [=attribution source/attribution scope limit=] to null.
+        1. Set |source|'s [=attribution source/attribution scopes=] to a new [=set=].
+    1. If the result of running [=check if source has incompatible attribution scope fields=] with |source| and |pendingSource| is true:
```

Comment:
We should also delete the sources without attribution scopes if pending source has attribution scopes.